### PR TITLE
dvc: refactor config

### PR DESCRIPTION
**Pull request by Suor**
Advantages:
- read like a dict
- modify as a dict for testing or manipulation purposes
- wrap modifications into context manager to save them:
    ```python
    with config.edit("local") as conf:
        conf["core"]["remote"] = "new-default"

Design decisions:
- dropped string constants
- remote sections are parsed into a simple subdict upon load
- remote config schema is now defined by url scheme
- relative paths are handled transparently upon load/save
- user config and remote manipulations are moved to commands
- same dev manipulations are done by simply modifying a config dict

### CODE REVIEW

**Comment by efiop** [comment]
@Suor Looks amazing! Please check the tests though, looks like slight hickup with schemas for some remotes.

**Comment by Suor** [comment]
@efiop I see two issues: 
- setting `type` for hdfs remote fails, which makes sense, tests need to be adjusted
- windows paths not being recognized sometimes by validation

I will fix these tomorrow.

**Comment by efiop** [comment]
@Suor Please rebase.

**Comment by Suor** [comment]
@efiop rebased

File: dvc/remote/s3.py [hunk 1]
```diff
@@ -26,35 +26,33 @@ class RemoteS3(RemoteBASE):
...
         else:
             self.list_objects_api = "list_objects_v2"
 
-        self.use_ssl = config.get(Config.SECTION_AWS_USE_SSL, True)
+        self.use_ssl = config.get("use_ssl", True)
 
         self.extra_args = {}
 
-        self.sse = config.get(Config.SECTION_AWS_SSE, "")
+        self.sse = config.get("sse")
```
Comment (pared) [inline dvc/remote/s3.py]: Shouldn't here be `config.get("sse","")`? And if not, why do we do `config.get("acl","")`?

File: dvc/remote/local.py [hunk 2]
```diff
@@ -46,22 +36,16 @@ class RemoteLOCAL(RemoteBASE):
...
         self._file_mode, self._dir_mode = self.SHARED_MODE_MAP[shared]
 
         if self.protected:
             # cache files are set to be read-only for everyone
             self._file_mode = stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH
 
-        cache_dir = config.get(Config.SECTION_REMOTE_URL)
-
-        if cache_dir is not None and not os.path.isabs(cache_dir):
-            cwd = config[Config.PRIVATE_CWD]
```
Comment (efiop) [inline dvc/remote/local.py]: @Suor You might've lost `_cwd` logic, which was resolving relative paths based on config location. Or am I missing something here?

Comment (shcheklein) [reply dvc/remote/local.py]: I thought that we have test for this 🤔

File: dvc/remote/s3.py (hunk 1 above)
Comment (Suor) [reply dvc/remote/s3.py]: Doesn't matter really until it's boolean false it won't be used anyway.

File: dvc/remote/local.py (hunk 2 above)
Comment (Suor) [reply dvc/remote/local.py]: This is not needed anymore:

> relative paths are handled transparently upon load/save

Comment (efiop) [reply dvc/remote/local.py]: @Suor Could you please refer me to some part of the code?

File: dvc/remote/s3.py (hunk 1 above)
Comment (Suor) [reply dvc/remote/s3.py]: Removed `""` for consistency.

File: dvc/remote/local.py (hunk 2 above)
Comment (efiop) [reply dvc/remote/local.py]: Nevermind, I see it now.

**Comment by efiop** [comment]
@Suor Please check the DS and CC reports, there are some good points there.

File: dvc/config.py [hunk 3]
```diff
@@ -300,250 +261,134 @@ def init(dvc_dir):
...
+        return conf
 
-    def load(self):
-        """Loads config from all the config files.
+    @staticmethod
+    def _load_paths(conf, filename):
+        def resolve(path):
+            if os.path.isabs(path) or re.match(r"\w+://", path):
+                return path
+            return RelPath(os.path.join(cwd, path))
```
Comment (efiop) [inline dvc/config.py]: cwd is defined after this helper in which it is used. This is error-prone.

Comment (Suor) [reply dvc/config.py]: Why?

Comment (efiop) [reply dvc/config.py]: This is minor. Nothing functionally wrong about it because of the same namespace, but causes a mental hickup.

**Comment by Suor** [comment]
@efiop I actually looked them through a couple of days ago) Most of them not introduced by this PR, from the rest:
- `ByUrl` complexity [here](https://codeclimate.com/github/iterative/dvc/pull/3298) - splitting it is not a good idea
- not calling `dict.__init__()` [here](https://deepsource.io/gh/iterative/dvc/run/cadeccf3-9db0-4ab4-8882-658d7061de90/python/PYL-W0231/) - `Config` has very different constructor signature.
- `self` in `.validate()` [here](https://deepsource.io/gh/iterative/dvc/run/cadeccf3-9db0-4ab4-8882-658d7061de90/python/PYL-R0201/) and `if/elif` [here](https://deepsource.io/gh/iterative/dvc/run/cadeccf3-9db0-4ab4-8882-658d7061de90/python/PYL-R1705/) just nitpicks, can't cause any issues
- `get_dir()` [here](https://deepsource.io/gh/iterative/dvc/run/cadeccf3-9db0-4ab4-8882-658d7061de90/python/PYL-R1710/) protected by an assert.

While some stylistic changes can be done, that will bring no value. Diving into DS was a waste of time again.

File: dvc/config.py (hunk 3 above)
Comment (Suor) [reply dvc/config.py]: Defining subfunction in the middle causes mental hiccup too)

Comment (Suor) [reply dvc/config.py]: I can move it if that worries you.

**Comment by efiop** [comment]
@Suor 

> ByUrl complexity here - splitting it is not a good idea


:+1: 

> not calling dict.__init__() here - Config has very different constructor signature.

:+1: 

> self in .validate() here and if/elif here just nitpicks, can't cause any issues

These are worth fixing, it affects the code style, we avoid such things in other places.

> get_dir() here protected by an assert.

:+1: 

> While some stylistic changes can be done, that will bring no value. Diving into DS was a waste of time again.

Style is not a waste of time for DVC, please fix those few issues left.

**Comment by efiop** [comment]
Thanks @Suor ! :pray:

**Pull request merged**